blockchain/thresholdstate.go

```diff
@@ -465,6 +465,82 @@ func (b *BlockChain) deploymentState(prevNode *blockNode, version uint32, deploy
 	return invalidState, DeploymentError(deploymentID)
 }
 
+// thresholdStatelastChanged returns the last node before the current state of
+// consensus deployment agenda activated.
+func (b *BlockChain) thresholdStateLastChanged(version uint32, prevNode *blockNode, checker thresholdConditionChecker, cache *thresholdStateCache) (*blockNode, error) {
+	// The threshold state for the window that contains the genesis block is
+	// defined by definition.
+	confirmationWindow := int64(checker.RuleChangeActivationInterval())
+	svh := checker.StakeValidationHeight()
+	if prevNode == nil || prevNode.height+1 < svh+confirmationWindow {
+		return b.bestChain.Genesis(), nil
+	}
+
+	curState, err := b.thresholdState(version, prevNode, checker, cache)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get the ancestor that is the last block of the previous confirmation
+	// window in order to get its threshold state.  This can be done because
+	// the state is the same for all blocks within a given window.
+	wantHeight := calcWantHeight(svh, confirmationWindow, prevNode.height+1)
+	prevNode = prevNode.Ancestor(wantHeight)
+	for prevNode != nil {
+		state, err := b.thresholdState(version, prevNode, checker, cache)
+		if err != nil {
+			return nil, err
+		}
+
+		if state.State != curState.State {
+			return prevNode, nil
+		}
+
+		// Get the ancestor that is the last block of the previous
+		// confirmation window.
+		prevNode = prevNode.RelativeAncestor(confirmationWindow)
+	}
+
+	return b.bestChain.Genesis(), nil
+}
+
+// StateActivationHeight returns the height of the first block where the
+// current state of consensus deployment agenda activated.
+func (b *BlockChain) StateActivationHeight(hash *chainhash.Hash, version uint32, deploymentID string) (int64, error) {
+	// Fetch the block node of the provided hash.
+	node := b.index.LookupNode(hash)
+	if node == nil || !b.index.NodeStatus(node).KnownValid() {
+		return 0, HashError(hash.String())
+	}
+
+	// Fetch the treshold state cache for the provided deployment id as well as
+	// the condition checker.
+	var cache *thresholdStateCache
+	var checker thresholdConditionChecker
+	for k := range b.chainParams.Deployments[version] {
+		if b.chainParams.Deployments[version][k].Vote.Id == deploymentID {
+			checker = deploymentChecker{
+				deployment: &b.chainParams.Deployments[version][k],
+				chain:      b,
+			}
+			cache = &b.deploymentCaches[version][k]
+			break
+		}
+	}
+
+	if cache == nil {
+		return 0, fmt.Errorf("Threshold state cache for agenda with "+
+			"deployment id (%s) not found", deploymentID)
+	}
+
+	stateChangeNode, err := b.thresholdStateLastChanged(version, node, checker, cache)
+	if err != nil {
+		return 0, err
+	}
+
+	return stateChangeNode.height + 1, err
```

Comment:
This is incorrect.  The function which determines the node in question should be returning the correct node.  This would return 2 in the case the genesis block node was returning which is clearly not correct!